diff --git a/mlir/lib/Bytecode/Reader/BytecodeReader.cpp b/mlir/lib/Bytecode/Reader/BytecodeReader.cpp
index 8ba64096fbb0..3008cd55eb02 100644
--- a/mlir/lib/Bytecode/Reader/BytecodeReader.cpp
+++ b/mlir/lib/Bytecode/Reader/BytecodeReader.cpp
@@ -1350,7 +1350,7 @@ LogicalResult AttrTypeReader::initialize(
 }

 template <typename T>
-T AttrTypeReader::resolveEntry(SmallVectorImpl<Entry<T>> &entries, size_t index,
+T AttrTypeReader::resolveEntry(SmallVectorImpl<Entry<T>> &entries, uint64_t index,
                                StringRef entryType, uint64_t depth) {
   bool oldResolving = resolving;
   resolving = true;
